utils.go

```diff
@@ -0,0 +1,168 @@
+		if err != nil {
+			return nil, err
+		}
+		// Cache the GetChanInfo result since it might be useful
+		hopHintChans[i] = chanInfo
+
+		//Skip if channel can't forward payment
+		if channel.RemoteBalance < amtMSat {
+			log.Infof(
+				"Skipping ChannelID: %v for hints as remote balance (%v sats) is insufficient"+
+					"appears to be private",
+				channel.ChannelID, channel.RemoteBalance,
+			)
+			continue
+		}
+		// If includeNodes is set, we'll only add channels with peers in
+		// include_node. This is done to respect the last_hop parameter
+		if len(includeNodes) > 0 {
+			if _, ok := includeNodes[channel.PubKeyBytes]; !ok {
+				continue
+			}
+		}
+
+		// We want to prevent leaking private nodes, which we define as
+		// nodes with only private channels
+		//
+		// GetNodeInfo will never return private channels, even if
+		// they're somehow known to us. If there are any channels
+		// returned, we can consider the node to be public.
+		nodeInfo, err := lnd.Client.GetNodeInfo(
+			ctx, channel.PubKeyBytes, true,
+		)
+		if err != nil {
+			return nil, err
+		}
+		if len(nodeInfo.Channels) == 0 {
+			log.Infof(
+				"Skipping ChannelID: %v for hints as peer (NodeID: %v)"+
+					"appears to be private",
+				channel.ChannelID, channel.PubKeyBytes.String(),
+			)
+			continue
+		}
+
+		nodeID, err := btcec.ParsePubKey(channel.PubKeyBytes[:], btcec.S256())
+		if err != nil {
+			return nil, err
+		}
+
+		// Now that we now this channel use usable, add it as a hop
+		// hint and the indexes we'll use later.
+		hopHints = append(hopHints, []zpay32.HopHint{{
+			NodeID:      nodeID,
+			ChannelID:   channel.ChannelID,
+			FeeBaseMSat: uint32(chanInfo.Node2Policy.FeeBaseMsat),
+			FeeProportionalMillionths: uint32(
+				chanInfo.Node2Policy.FeeRateMilliMsat,
+			),
+			CLTVExpiryDelta: uint16(chanInfo.Node2Policy.TimeLockDelta),
+		}})
+
+		totalHintBandwidth += channel.RemoteBalance
+	}
+
+	// If we have enough hop hints at this point, then we'll exit early.
+	// Otherwise, we'll continue to add more that may help out mpp users.
+	if len(hopHints) >= numMaxHophints {
+		return hopHints, nil
+	}
+
+	// In this second pass we'll add channels, and we'll either stop when
+	// we have 20 hop hints, we've run through all the available channels,
+	// or if the sum of available bandwidth in the routing hints exceeds 2x
+	// the payment amount. We do 2x here to account for a margin of error
+	// if some of the selected channels no longer become operable.
+	hopHintFactor := btcutil.Amount(lnwire.MilliSatoshi(2))
+
+	for i := 0; i < len(openChannels); i++ {
+		// If we hit either of our early termination conditions, then
+		// we'll break the loop here.
+		if totalHintBandwidth > amtMSat*hopHintFactor ||
+			len(hopHints) >= numMaxHophints {
+
+			break
+		}
+
+		// Skip the channel if we already selected it.
+		if _, ok := hopHintChans[i]; ok {
+			continue
+		}
+		channel := openChannels[i]
+		chanInfo := hopHintChans[i]
```

Comment:
This isn't working (see where we should add to `hopHintChans`). It's only containing the channel info because we add all channels to the `hopHintChans` (which is why we never reach this far).